### PR TITLE
Azure monitor Logs: Optimize data fetching in resource picker

### LINKED ISSUE #35643: AzureMonitor: Add error handling to ResourcePicker
**What would you like to be added**:

Catch HTTP errors in the resource picker when:
 - fetching initial data
 - fetching resources in resource groups

and display the error in the UI consistent with the rest of the data source.

### PR DESCRIPTION
**Pull request by sunker**
**What this PR does / why we need it**:

Some users have reported that the resource picker in Azure Monitor logs crashes their browser while attempting to open it. This seems to happen when the account has a large number of resources. 

This PR aims to optimize the way data is fetched from the Azure resource graph api while opening the resource picker modal. 

This is what happened when the resource picker was opened before this PR: 
1. Load all subscriptions and resource groups in one request
3. Load resources (if one was previously selected)

Change in this PR:
1. Load all subscriptions in one request
2. Load resource groups for one subscription (if one was previously selected)
3. Load resources for one resource group (if one was previously selected)

From a user standpoint, the only visual change will be that in case no resource was previously selected, only the first level (subscriptions) will be displayed after the resource picker i opened. Previously, two levels (subscriptions and resource groups) were being displayed. 

Using the new approach, we only fetch data that will be displayed in the UI. If user expands a subscription or resource group item in the tree, it will issue a new api call (unless that item had been expanded before since they're being cached). 

This new approach does not rule out the risk of getting out memory exceptions in the case a subscription has thousands of resource groups or a resource group has thousands of resources. However, it seems like this is the approach that was chosen for the scope picker in the Azure Portal so it should work here too. If not, we can consider implementing [pagination support](https://github.com/grafana/grafana/discussions/41809) in the UI. 

The resource picker needed error handling for meta data requests, so that is also added in this pr. 

**Which issue(s) this PR fixes**:
Related to https://github.com/grafana/grafana/issues/42106 https://github.com/grafana/grafana/issues/28016
Fixes #35643

Will fix escalation https://github.com/grafana/support-escalations/issues/1405

**Special notes for your reviewer**:
~~This PR is ready for review, but it's still in draft mode because I don't want to merge it before I receive feedback from a customer.~~ 

Future considerations: 
It's starting to become painful to manage state in the resource picker. If we add more features to it (like multi-select), we should consider using redux or similar.

### CODE REVIEW

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.tsx [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+          setLoadingStatus('Started');
+          let resources = await resourcePickerData.getSubscriptions();
+          if (!internalSelected) {
+            setAzureRows(resources);
+            setLoadingStatus('Done');
+            return;
+          }
+
+          const parsedURI = parseResourceURI(internalSelected ?? '');
+          const resourceGroupURI = `/subscriptions/${parsedURI?.subscriptionID}/resourceGroups/${parsedURI?.resourceGroup}`;
```
Comment (andresmgot) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.tsx:56]: thinking out loud but what if `internalSelected` is a subscription?

Comment (andresmgot) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.tsx:56]: I see that it won't enter in the other `if`s. Nit but for readability I would write:

```
const parsedURI = parseResourceURI(internalSelected ?? '');
if (parsedURI?.resourceGroup) {
             const resourceGroupURI = `/subscriptions/${parsedURI?.subscriptionID}/resourceGroups/${parsedURI?.resourceGroup}`;
             ....
```

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts [hunk 2]
```diff
@@ -101,14 +141,13 @@ export default class ResourcePickerData extends DataSourceWithBackend<AzureMonit
...
+    resourcecontainers
+    | where type == "microsoft.resources/subscriptions"
+    | where id =~ "${subscriptionURI}"
+    | project subscriptionName=name, subscriptionId
 
-        | join kind=leftouter (
-          resourcecontainers
-            | where type == "microsoft.resources/subscriptions/resourcegroups"
+    | join kind=leftouter (
+      resourcecontainers            | where type == "microsoft.resources/subscriptions/resourcegroups"
```
Comment (andresmgot) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts:150]: missing new line here

**Review by andresmgot — APPROVED** [review]
This is great! I only found a couple of nits and I tested this locally so for me it's a good to go.

Buut, what if we also handle "pagination" here? Meaning, if a request returns a "next page" we just concat the result. This is what we do now for Metrics so if we are going to use the resource picker there we should have that feature. It does not need to be in this PR though, what do you think?

**Comment by sunker** [comment]
Thanks for feedback!

> Buut, what if we also handle "pagination" here? Meaning, if a request returns a "next page" we just concat the result. This is what we do now for Metrics so if we are going to use the resource picker there we should have that feature. It does not need to be in this PR though, what do you think?

Do you mean for the log queries or for the resource picker? I believe all resource picker queries except for loading resources under a resource group already support pagination.

**Pull request merged**